### ISSUE TITLE
Object: 'invokeWithExtensions' $argument should be call-by-reference to be consistent with 'extend'

### ISSUE THREAD

**Issue by peda** [silverstripe/silverstripe-framework#3745]
In the Object class currently the `$argument` of `invoke withExtensions` is a call-by-value:

``` php
public function invokeWithExtensions($method, $argument = null)
```

whereas the arguments are call-by-reference for the method `extend`:

``` php
public function extend($method, &$a1=null, &$a2=null, &$a3=null, &$a4=null, &$a5=null, &$a6=null, &$a7=null) 
```

In order to be consistent it would be great if the `$argument` argument of `invokeWithExtensions` could be changed to call by-reference as well. This inconsistency already lead to mistakes even in silvertripe-labs modules (https://github.com/silverstripe-labs/silverstripe-googlesitemaps/issues/78)

**Comment by patricknelson** [comment]
:+1: I think I'll do a quick PR against this since it's pretty annoying.

**Comment by tractorcow** [comment]
Too late I've already done it. It's getting PR up soon. :D

**Comment by patricknelson** [comment]
Wait... hey! So -- you put that up against `3` right? And you were fine with the mild API change, that is: Forcing `$arg1` to effectively be a pass by reference making calls like `$obj->invokeWithExtensions('method', 'string constant')` illegal)?

**Comment by tractorcow** [comment]
It's against master at the moment, and is a part of a larger piece of work.

https://github.com/open-sausages/silverstripe-framework/commit/697d57bdba55db87a40b6c64732c6e8f0df4d02a#diff-82f6624f9301c842e9bd32d95d0f4be4R380

No I'm not fine with the API change... method signature changes are (nearly) always semver breaking.

**Comment by patricknelson** [comment]
Yeah that's what I thought. Only after I dove into this did I realize that switching to pass-by-reference would likely break calls which were using literals and not variables :disappointed:  This makes my work **much** more frustrating since I refuse to modify core unless I know it'll make it into `3` so, instead, I'll try to approach a fix via PR against https://github.com/silverstripe-labs/silverstripe-googlesitemaps/issues/78 instead.

**Comment by patricknelson** [comment]
Thanks for the head's up, by the way! Saved me a little bit of time at least!

**Comment by tractorcow** [comment]
See https://github.com/silverstripe/silverstripe-framework/pull/5094
